### PR TITLE
[WIP] sundials interface + example

### PR DESCRIPTION
**Pull request by asartori86**

### CODE REVIEW

**Comment by asartori86** [comment]
guys, could you take a look at the `heat_ida example` in the directory `examples` I just split the `test/heat_ida_01.cc` file (which works well) but I got strange error of compilation. 

thanks

File: CMakeLists.txt [hunk 1]
```diff
@@ -73,37 +89,50 @@ FOREACH(_build_type ${_d2_build_types})
...
-	SET(TEST_LIBRARIES_${_BUILD_TYPE} ${_lib})
-	DEAL_II_SETUP_TARGET(${_lib} ${_BUILD_TYPE})
+  SET_PROPERTY(TARGET ${_lib} APPEND PROPERTY COMPILE_DEFINITIONS
+    DEAL_II_SAK_GIT_BRANCH="${GIT_BRANCH}${_postfix}"
+    DEAL_II_SAK_GIT_REVISION="${GIT_REVISION}${_postfix}"
+    DEAL_II_SAK_GIT_SHORTREV="${GIT_SHORTREV}${_postfix}"
+  )
+
+  # Compile an executable only if required by the user
+  IF(DEAL_II_SAK_COMPILE_EXAMPLES)
```
Comment (luca-heltai) [inline CMakeLists.txt]: This option should disappear. The examples should only be _installed_, not compiled. Of course we should test that each example compiles, but that's an entirely different story...

File: CMakeLists.txt [hunk 2]
```diff
@@ -73,37 +89,50 @@ FOREACH(_build_type ${_d2_build_types})
...
+         FOREACH(_main ${_applications_source_files})
+    # Find out the the name of the executable
+        GET_FILENAME_COMPONENT(_target_a ${_main} NAME_WE)
+    SET(_target "${_target_a}${${_build_type}_postfix}")
+    MESSAGE("-- Configuring executable ${_target}")
+
+        ADD_EXECUTABLE(${_target} ${_main})
```
Comment (luca-heltai) [inline CMakeLists.txt]: I'd remove from my previous comment to here...

File: examples/heat_equation/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,121 @@
+##
+# CMake script for large project, with one executable
+# and a library of objects, optionally
```
Comment (luca-heltai) [inline examples/heat_equation/CMakeLists.txt:3]: Not optionally, and not on dealii-dak

File: examples/heat_equation/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+)
+
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-pedantic")
+DEAL_II_INITIALIZE_CACHED_VARIABLES()
+PROJECT(${TARGET})
+
+
+SET(DEAL_II_SAK "$ENV{DEAL_II_SAK}" CACHE PATH "Path to dealii-sak")
+
+# We one library and one target for each type of deal.II library
```
Comment (luca-heltai) [inline examples/heat_equation/CMakeLists.txt:54]: we _build_

File: examples/heat_equation/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  HINTS ${deal.II_DIR} ${DEAL_II_DIR} ../ ../../ $ENV{DEAL_II_DIR}
+)
+
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-pedantic")
+DEAL_II_INITIALIZE_CACHED_VARIABLES()
+PROJECT(${TARGET})
+
+
+SET(DEAL_II_SAK "$ENV{DEAL_II_SAK}" CACHE PATH "Path to dealii-sak")
+
```
Comment (luca-heltai) [inline examples/heat_equation/CMakeLists.txt:53]: Add an option for the testing:

```
OPTION(ENABLE_TARGET_TESTING "Enable testing for this target." OFF)
```

File: examples/heat_equation/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	    DEAL_II_SETUP_TARGET(${_exe} ${_BUILD_TYPE})
+	ENDIF()
+	SET(TEST_LIBRARIES_${_BUILD_TYPE} ${_lib})
+	DEAL_II_SETUP_TARGET(${_lib} ${_BUILD_TYPE})
+    ENDIF()
+ENDFOREACH()
+
+INCLUDE_DIRECTORIES(${DEAL_II_SAK}/include)
+
+if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/tests)
```
Comment (luca-heltai) [inline examples/heat_equation/CMakeLists.txt:114]:
```
IF(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/tests AND ENABLE_TARGET_TESTING)
```

File: examples/heat_equation/include/heat_ida.h [hunk 7]
```diff
@@ -0,0 +1,200 @@
+#include <deal.II/lac/sparsity_tools.h>
```
Comment (luca-heltai) [inline examples/heat_equation/include/heat_ida.h]: Add 

``` C++
#ifndef __heat__ida__h
#define __heat__ida__h
....
#endif
```

guard for header files.

File: examples/heat_equation/include/heat_ida.h [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+#include "sundials_interface.h"
+#include "dae_time_integrator.h"
+#include "parameter_acceptor.h"
+#include "parsed_grid_generator.h"
+#include "parsed_finite_element.h"
+#include "error_handler.h"
+#include "parsed_function.h"
+#include "parsed_data_out.h"
+#include "parsed_dirichlet_bcs.h"
```
Comment (luca-heltai) [inline examples/heat_equation/include/heat_ida.h:26]: Remove all unnecessary headers. Only those which are required for the h file should be here, the others should be in the cc file.

File: cmake/config/Config.cmake.in [hunk 9]
```diff
@@ -0,0 +1,188 @@
+## ---------------------------------------------------------------------
+##
+## Copyright (C) 2012 - 2015 by the deal.II authors
+##
+## This file is part of the deal.II library.
```
Comment (luca-heltai) [inline cmake/config/Config.cmake.in:5]: deal2ltkit?

File: cmake/config/Config.cmake.in [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+## version 2.1 of the License, or (at your option) any later version.
+## The full text of the license can be found in the file LICENSE at
+## the top level of the deal.II distribution.
+##
+## ---------------------------------------------------------------------
+
+
+########################################################################
+##                                                                    ##
+##               The deal.II project configuration file               ##
```
Comment (luca-heltai) [inline cmake/config/Config.cmake.in:19]: same here

**Comment by luca-heltai** [comment]
Can you rebase?

**Comment by luca-heltai** [comment]
If you want me to merge, remove the [WIP]. ;)

**Comment by luca-heltai** [comment]
Also, please squash away the commits you know are not working...

**Pull request closed without merging**